### ISSUE TITLE
Add ligatures

### ISSUE THREAD

**Issue by larsenwork** [larsenwork/monoid#25]
Like e.g. https://github.com/tonsky/FiraCode

I'm still trying to figure out how best to solve it technically though

**Comment by larsenwork** [comment]
So I think I've figured out how to do some of it. I'm trying to keep this a strictly monospaced font so hack job like combining two characters in to one wider character isn't kosher. Instead I'm thinking about using contextual substitution like this
```
feature calt {
  sub less hyphen' by ligaarrowleft;
} calt;
feature calt {
  sub less' ligaarrowleft by space;
} calt;

```
so that all glyphs still have the same width. 

![screen shot 2015-06-06 at 07 53 14](https://cloud.githubusercontent.com/assets/3859425/8018523/30f846cc-0c21-11e5-8d0b-27193df51175.png)

**Comment by larsenwork** [comment]
![monoidvsroboto](https://cloud.githubusercontent.com/assets/3859425/8146246/211b4ea8-122e-11e5-9e9a-ce9dfb9f394d.png)

I've added these so far but having some issues with "calt" and Atom (editor I'm primarily using)

**Comment by glebd** [comment]
I wonder if this is going too work in Visual Studio on Windows. Pragmata Pro and Hasklig ligatures seem to work there. In case they work, it would be great to have =>, // and /// for C# in addition to the ones above. Also, for F# (which is cross-platform) |>, <|, <||, <||| (* *), (| |), ::, :>, :=, <<<, >>>, [| |], [< >] etc. would be great. The full list: https://msdn.microsoft.com/en-us/library/dd233228.aspx

**Comment by larsenwork** [comment]
@glebd can you check the latest version and see if it works on windows with these ligatures https://github.com/larsenwork/monoid/issues/25#issuecomment-111757957

**Comment by larsenwork** [comment]
@glebd should => be an arrow like this: ⇒ or what is it used for? I have added // and /// (not pushed yet)

**Comment by glebd** [comment]
Yes, I think so: => turns into ⇒ across 2 character widths. Thanks!

**Comment by larsenwork** [comment]
@glebd cheers, did you get to test it on Windows?

**Comment by glebd** [comment]
Not yet, will do next week

**Comment by larsenwork** [comment]
:+1:

**Comment by larsenwork** [comment]
![screen shot 2015-07-19 at 22 26 04](https://cloud.githubusercontent.com/assets/3859425/8767727/100bd4ee-2e66-11e5-98a5-73c4b38ba978.png)

**Comment by larsenwork** [comment]
Compared to 
![screen shot 2015-07-19 at 22 33 00](https://cloud.githubusercontent.com/assets/3859425/8767731/21a03f74-2e66-11e5-8720-5a2d11860b1c.png)

**Comment by glebd** [comment]
Niiice

**Comment by larsenwork** [comment]
@glebd I don't know haskell but is this something I should fix (comment about >→) https://news.ycombinator.com/item?id=9718099 

And are there other similar issues?

**Comment by mo-seph** [comment]
These are lovely!

I'm writing in a language where => and <= are both arrows, and at the moment it's a bit confusing as:
* <= gets turned into less than/equal to
* => doesn't seem to be ligatured
* <- and <-- are correctly ligatured into arrows

Unfortunately, some of this is language and context dependent. For me there are times when <= is an arrow, and times when it is less that or equal to.

Could the download page (which is great, btw) allow a choice of ligatures as well as just on and off?

**Comment by larsenwork** [comment]
@mo-seph cheers, 
=> should be a ligature. What editor + OS are you using?
![screen shot 2015-07-24 at 10 40 30](https://cloud.githubusercontent.com/assets/3859425/8870724/6c8cfe4a-31f0-11e5-9497-39cc5a19d2d7.png)

<= vs ≤ could be a choice, the easiest (and best thing) would be to add it as an stylistic alternate, but that only works if your editor allows you to add your own stylesheet.

It could also be a downloadable choice but problem is we're currently generating all possible versions (current 3200) so adding it as downloadable choice would require it replacing one of the existing options.

**Comment by larsenwork** [comment]
What language uses <= as arrow?

**Comment by mo-seph** [comment]
An obscure one ;) It's called LCC, and represents messages sent to ( => ) and from ( <= ) agents
(https://bitbucket.org/mo_seph/scalsc/wiki/Representing%20Interactions)

Here's a screenshot, using Eclipse on OSX 10.10
<img width="636" alt="screen shot 2015-07-24 at 09 50 19" src="https://cloud.githubusercontent.com/assets/509387/8870944/bd879712-31e9-11e5-8866-fe2f6cb757fc.png">

The <= on line 7 has been changed to leq as expected, the => on line 8 is un-liga'd, and the <- on line 9 is great ;)

(BTW, these arrows are clearly an edge case, and I suspect your approach works for 99% of the time ;) )

**Comment by larsenwork** [comment]
@mo-seph => should be a ligature. What editor + OS are you using?

**Comment by mo-seph** [comment]
Eclipse, OSX 10.10.4

**Comment by flaviotruzzi** [comment]
Ligatures are not working here :( (IntelliJ  OSX 10.10.4)

**Comment by chase** [comment]
@larsenwork It seems like Java doesn't support ligatures, at least not on OS X. Eclipse and IntelliJ are both Java-based.

**Comment by mikebronner** [comment]
Crud, I guess that holds true for PHPStorm 9 as well (built on Java).

**Comment by chase** [comment]
@mikebronner Since it is a derivative of IntelliJ, I'd say that's safe to assume.

**Comment by rkurbatov** [comment]
@mikebronner @chase Eclipse is supporting ligatures, just checked. IntelliJ doesn't - they've changed rendering engine. They are working at this. SublimeText too.

**Comment by chase** [comment]
@rkurbatov Thanks for looking into it!

**Comment by larsenwork** [comment]
Please add your findings to the readme #67

**Comment by larsenwork** [comment]
@glebd About the // ligature: would it make sense to only activate it if it's preceded/followed by a "space" ?

I want to turn it off for at least http:// which currently look a bit odd with the ligature

**Comment by glebd** [comment]
Not really, as C-like languages use // for comments which may start at the beginning of the line.

**Comment by larsenwork** [comment]
@glebd space after?

**Comment by larsenwork** [comment]
I've never us

**Comment by glebd** [comment]
Hmm, some people don't leave space after //comments.

**Comment by larsenwork** [comment]
Hmm...they should, looks messy:) Well I can easily "filter out" http:// so I'll just take that approach instead.

**Comment by glebd** [comment]
Cool.

**Comment by plgruener** [comment]
suggested/requested ligatures (mostly from Java and Groovy, see full list [here](http://docs.groovy-lang.org/latest/html/documentation/#_operator_precedence) )
++ (maybe same as with == and --)
operators += -= \*= /= %= \** \**= =~ ==~
<=> "Spaceship-operator", currently this is broken into <= and >
for comments (without the spaces) / * , / * * ,  */ (similar to //)
also cool would be ?:  ?.  !! *. ..<

**Comment by larsenwork** [comment]
One could do the same as == and -- for ++ but not sure. Plus signs have a vertical line making it easy to tell them apart so not sure the ligature is necessary
![screen shot 2015-07-27 at 12 26 12](https://cloud.githubusercontent.com/assets/3859425/8903787/b4db9b40-345a-11e5-9d1d-ece1faf7226e.png)
Spaceship: We could go crazy and do something like this
![screen shot 2015-07-27 at 12 25 33](https://cloud.githubusercontent.com/assets/3859425/8903830/0c280d34-345b-11e5-80f3-ec97bf12e255.png)
but I'm more inclined to do something like this
![screen shot 2015-07-27 at 12 28 34](https://cloud.githubusercontent.com/assets/3859425/8903835/14bbcdb4-345b-11e5-92d5-fea5dc1d747c.png)
As for the operators we could do narrower +-= signs but not sure it's needed
![screen shot 2015-07-27 at 12 32 23](https://cloud.githubusercontent.com/assets/3859425/8903936/a3bee154-345b-11e5-81d3-6685dffdac53.png)

**Comment by plgruener** [comment]
Spaceship: that indeed looks crazy somehow :D But I'd also go with the (sane) second version.
++ and ** : right, probably not needed. But it looked slightly strange because all others (==, --) have these little gaps.

Same goes for *= %= etc. Not really necessary, but I think a little gap (narrower = signs) would make it prettier (just my personal opinion, you're the font expert ;)

**Comment by larsenwork** [comment]
Been looking at it and I might just change the default =+- to the narrower versions

**Comment by larsenwork** [comment]
So it'd be like so:
![screen shot 2015-07-27 at 13 49 54](https://cloud.githubusercontent.com/assets/3859425/8905413/66c4ecf2-3466-11e5-8499-c7149b0b83e0.png)

**Comment by plgruener** [comment]
Well, that'd save you some work with the ligatures. However, I don't quite like that asymmetrical plus sign, looks a bit odd to me.
I don't know about your build process, but maybe you can offer an alternate version with narrow +-= signs, so some more people can test it (at different sizes).

**Comment by larsenwork** [comment]
Yeah...I actually like the narrower equalsign as default and then keeping the + and - unchanged but have the ligature for --

![screen shot 2015-07-27 at 14 52 53](https://cloud.githubusercontent.com/assets/3859425/8906498/753bfa2e-346f-11e5-8b35-eb56dd0ef7ee.png)
![screen shot 2015-07-27 at 14 54 27](https://cloud.githubusercontent.com/assets/3859425/8906501/796c3ee2-346f-11e5-9608-98929bc0bb7c.png)
![screen shot 2015-07-27 at 14 55 31](https://cloud.githubusercontent.com/assets/3859425/8906513/8b960076-346f-11e5-8700-5a4c4ad4493c.png)

**Comment by plgruener** [comment]
Yes, sounds like a good solution. Do the unligatured versions of >= => etc. still look good with the narrow equal-sign?

**Comment by larsenwork** [comment]
@plgruener I think so - they look like the spaceship (3rd pic): https://github.com/larsenwork/monoid/issues/25#issuecomment-125159686

**Comment by plgruener** [comment]
Ah right. Looks great. (Except the -= has some unfamiliar proportions, but one should get accustomed to it pretty fast.)

**Comment by larsenwork** [comment]
The - in -= could be a hybrid between - and the - in -- 
![screen shot 2015-07-27 at 16 30 26](https://cloud.githubusercontent.com/assets/3859425/8908518/d4de2f12-347c-11e5-9652-6621e7b8b12b.png)

**Comment by larsenwork** [comment]
Moving the first asterisk to the left and the second up and to the right to make more room for the comment
![screen shot 2015-07-28 at 01 24 55](https://cloud.githubusercontent.com/assets/3859425/8920274/84dc9364-34c7-11e5-9ca7-d78c688a5d10.png)

**Comment by larsenwork** [comment]
@plgruener what are `?: ?. !! *. ..<` used for? I need some context examples to see what ligatures (if any) will be fitting.

**Comment by plgruener** [comment]
Keep in mind one often formats comments like this:
```java
/**
 *
 */
```
So I don't know about left/right shifting the asterisk…

**Comment by larsenwork** [comment]
It will look a bit weird but the gain in /*Comment*/ cases should be worth it.

**Comment by plgruener** [comment]
Well, actually comments like /* ... */ in one line are very discouraged (at least in Java & co.). But we could try it first.

**Comment by larsenwork** [comment]
In css they're everywhere :)

**Comment by larsenwork** [comment]
I actually think it works ok:

![screen shot 2015-07-28 at 01 57 18](https://cloud.githubusercontent.com/assets/3859425/8920645/08c155ee-34cc-11e5-8443-424c8101dd9d.png)

**Comment by plgruener** [comment]
the `!!` is a bash alias for previous command, it stands alone. And there also seems to be others with ! and a following char, like `!?` or `!#' – but I doubt they're often used, so it's not urgent. (Although sometimes I comment things with "what's that for?!" ;) )

Examples for the others:

`def name = person?.name` (assign name only if person is not null)
`def fun = str.&toUpperCase` (used to store a method pointer)
`def makes = cars*.make` (apply make to all elements of the cars-collection)
`def range = 0..<5` numbers 0 to 5 exclusively (in contrast to `0..5` which are 0 to 5 inclusively)

Most of them look ok without ligatures, except for ..<

**Comment by mikebronner** [comment]
There are also tempting syntaxes that use `{!! some command here !!}` and `{{ some command here }}` and `{{{ some command here }}}`. Would these be suited for ligatures as well?

**Comment by larsenwork** [comment]
@plgruener 
![screen shot 2015-07-28 at 03 13 33](https://cloud.githubusercontent.com/assets/3859425/8921505/b2942fb0-34d6-11e5-8899-fd01557758b8.png)
![screen shot 2015-07-28 at 03 12 03](https://cloud.githubusercontent.com/assets/3859425/8921507/b2ba4efc-34d6-11e5-9eab-a9d2dc89434e.png)
![screen shot 2015-07-28 at 02 43 43](https://cloud.githubusercontent.com/assets/3859425/8921506/b2b9c3f6-34d6-11e5-8196-73ab518e6537.png)

Diacritics are almost done
![screen shot 2015-07-28 at 03 15 33](https://cloud.githubusercontent.com/assets/3859425/8921525/fbbb2608-34d6-11e5-94dc-bcd18e1b85e1.png)

@mikebronner I think they could be improved...stay tuned :smile:

**Comment by plgruener** [comment]
Hmm, the same would apply for any amount of opening/closing brackets. I mean ()))) is also not uncommon, as well as ({}) or something.
I understand your approach of defining new ligatures is to assign a special sequence of characters to a new one. This wouldn't be suited for "any amount of brackets", would it?

**Comment by plgruener** [comment]
@larsenwork subtle differences, but looks good to me. Did you alter the position of the ..< ?

**Comment by plgruener** [comment]
I was looking at Haskell-ligatures (not using it, but these folks seem to like ligatures), and I think the following need fixing:
`=:=` center colon between the equalsigns, `::` squeeze them a bit together,
`<== ==>` should render to long double arrows like `=>`,
`>>=  =<<` are broken by the existing <= ligatures

evtl also `<<< , >>> , -< , >- , -<< , >>-`

**Comment by larsenwork** [comment]
@plgruener 
any-amount: ligatures are only really suited for 2-3(-4) characters. I'm currently building an update that fixes e.g. &&& (currently broken)
..< problem was ".." got recognised first and then left the "<" hanging. 
=:= can be centred although the : will not be as sharp. Seems like a good tradeoff though.
<== ==> >>= =<< should be an easy fix

**Comment by larsenwork** [comment]
![screen shot 2015-07-29 at 00 29 32](https://cloud.githubusercontent.com/assets/3859425/8945580/eadf76ce-3588-11e5-8362-ec9d0526c5c5.png)

**Comment by larsenwork** [comment]
@Fequois I've done the same as with hyphens (make them narrower when more than one) so it's easier to see how many underscores there are

![screen shot 2015-07-29 at 16 38 48](https://cloud.githubusercontent.com/assets/3859425/8960460/54932ea8-3610-11e5-848b-211dfbc85bdb.png)

**Comment by larsenwork** [comment]
compared to 
![screen shot 2015-07-29 at 16 39 57](https://cloud.githubusercontent.com/assets/3859425/8960486/7af4311e-3610-11e5-8310-2c1ee7f8df32.png)

**Comment by larsenwork** [comment]
nah...can be better, two sec...

**Comment by Fequois** [comment]
Hahaha, love your workflow. <3

*waits for the next set of images*

**Comment by larsenwork** [comment]
![screen shot 2015-07-29 at 16 50 02](https://cloud.githubusercontent.com/assets/3859425/8960772/e0ef94bc-3611-11e5-9469-9a8626a10dbb.png)

**Comment by plgruener** [comment]
yep, looks clearer now. so this is for any number of underscores (>1) ?

**Comment by larsenwork** [comment]
@plgruener yep but n=2 is a special case
![screen shot 2015-07-29 at 16 52 39](https://cloud.githubusercontent.com/assets/3859425/8960865/50b8da1a-3612-11e5-8271-9ae149a0d194.png)

**Comment by Fequois** [comment]
@larsenwork I like it, too. Very nice.

**Comment by plgruener** [comment]
hmm, same maybe for ~~ ?

**Comment by larsenwork** [comment]
Thanks, up next is perfect shrug `¯\_(ツ)_/¯` ligature...just because...

**Comment by larsenwork** [comment]
@plgruener I think the ~ will get too condensed + it looks ok now because it's not "end-to-end" so it's easier to tell them apart if there are many

**Comment by larsenwork** [comment]
https://twitter.com/larsenwork/status/626424006174973952

**Comment by plgruener** [comment]
Yeah, priorities! :+1: 

@ ~~ I just wanted to throw in the idea.

**Comment by cabralRodrigo** [comment]
I don't know is this is the right issue but I'm using monoid to develop typescript and the ligatures for the operator `!==` seems a little odd.

![ts](https://cloud.githubusercontent.com/assets/5928832/8985659/da04cbde-36ae-11e5-92a9-06c28c98b262.png)

I don't know if is even possible to do something about it, but would be really nice if this could change :) 

BTW, awesome font to develop :)

**Comment by larsenwork** [comment]
It's how it's designed "notequal + type" "≠="

Any other ideas as to how it could/should look?

**Comment by cabralRodrigo** [comment]
I really don't have any ideas, maybe something like this:
![n](https://cloud.githubusercontent.com/assets/5928832/8985992/55fbc35e-36b0-11e5-81cd-ff050776fb7a.png)

But I can see that this can be a little confusing with the current `!=`.

And, if this is this way by design I can get used to it.

**Comment by larsenwork** [comment]
@cabralRodrigo yep, I deliberately didn't do like e.g. fira mono because I don't want there to be any confusion between !== and !=

**Comment by cabralRodrigo** [comment]
@larsenwork Thanks for your time anyway :)

**Comment by luben** [comment]
In gEdit ">==>" is rendered like a " ==>" (with leading space) : not that I know of programming language that uses it.

**Comment by luben** [comment]
Also "foo <- bar" is rendered as "foo  - bar"

**Comment by luben** [comment]
Correttion on the previous comment, regarding "foo <- bar". It renders incorrectly with the Retina face but renders correctly with the Regular one. ">==>" renders incorrecly in both of them

**Comment by larsenwork** [comment]
@luben cheers, "<-" is fixed in next update 
Because of my approach (using contextual alternates and not conventional ligatures (read why in the medium articles)) random strings of "<", "=" and ">" will render with errors...not sure how to fix this. Other than to include specific solutions for those combinations used by programming languages (e.g. <==>)

**Comment by blixt** [comment]
@larsenwork The `!==` ligature is very confusing to me and was the one thing that stood out to me as something that makes code less readable. To me, and I'm sure many others, it reads pretty much like a broken form of regular not equals.

Semantically, identity equals is a three line equals sign (≡) and with a strikethrough it would be equivalent to `!==`. I would do just like Fira Code does it, which is immediately clear to me:

<img width="75" alt="screen shot 2015-10-29 at 9 41 08 pm" src="https://cloud.githubusercontent.com/assets/158591/10836644/50be3516-7e86-11e5-9649-36832defadc6.png">

If keeping the current one is something you feel strongly about, could the more semantically correct version perhaps be available as an optional font feature?

**Comment by in-in** [comment]
I agree with @blixt. The first time I saw that `!==` ligature, I thought it is broken. It is look weird in such perfect font. Double m also confusing me.

**Comment by larsenwork** [comment]
@blixt a good idea, I will look into it this christmas :+1:

**Comment by larsenwork** [comment]
@blixt I have a problem with the "==" ligature though...it's aesthetically doing what e.g. Fira Code does and merge the two equal signs but then it's only by length that you can distinguish it from a single equal sign...not sure what's the optimal solution

**Comment by larsenwork** [comment]
![screen shot 2015-12-10 at 22 03 45](https://cloud.githubusercontent.com/assets/3859425/11728406/ec665010-9f89-11e5-9086-4b94a48c848f.png)

**Comment by in-in** [comment]
You're right, it looks a bit alike.
I tried your sample in editor (Fira Code).
![2](https://cloud.githubusercontent.com/assets/8797432/11729627/3c9c4b7c-9fa1-11e5-9614-3a73c687754b.jpg)
You can see different height between lines.

**Comment by marnen** [comment]
@larsenwork:
> I'm trying to keep this a strictly monospaced font so hack job like combining two characters in to one wider character isn't kosher. Instead I'm thinking about using contextual substitution like this

A belated note of thanks for this idea. It's what I used in https://github.com/marnen/borg-sans-mono.

**Comment by larsenwork** [comment]
@marnen cheers, I've recently changed my opinion about this a bit though so I think I'll make more ligatures that span multiple letters 🤷‍♂️ 😉

**Comment by marnen** [comment]
@larsenwork Does that mean that you're now doing something like this?
```fea
sub a' b by a_b.left_half
sub a_b.left_half b' by a_b.right_half
```

**Comment by larsenwork** [comment]
@marnen essentially yeah, because more and more editors and terminals support this kind of ligatures now (didn't do that when I created Monoid)

**Comment by marnen** [comment]
Right. I don't especially like the idea of the left and right halves being separate glyphs in the font, so I probably wouldn't do it that way unless there were some overwhelming advantage.

**Comment by larsenwork** [comment]
@marnen they actually aren't in Monoid either — if it's a ligature that spans three letters then the first two are substituted to empty letters and the third one has the entire ligature. But it's complicated to maintain so normal ligatures is obviously they way to go now that editors support it.

**Comment by marnen** [comment]
Does that work in a monospaced font, though? I wouldn't expect it to, at least not with proper caret positioning, hence the space trick that I use in Borg Sans.

**Comment by larsenwork** [comment]
that's how all my "ligatures" have been created so far but it's annoying having to substitute with 3 characters instead of just one.
